### PR TITLE
add asset signed url proxy logic for docker compose context

### PR DESCRIPTION
**Pull request by sergzak022**

### CODE REVIEW

File: local_dispatcher/local_dispatch.py [hunk 1]
```diff
@@ -19,12 +19,13 @@
 
 
 class Service():
-    def __init__(self, name, host, port, path_function, is_default=False):
+    def __init__(self, name, host, port, path_function, is_default=False, get_service_path=lambda path : path):
```
Comment (sergzak022) [inline local_dispatcher/local_dispatch.py:22]: `get_service_path` can be used to modify service path. For example, I use it to remove `proxy_to_minio` portion from the url before sending it to minio.

File: local_dispatcher/local_dispatch.py [hunk 2]
```diff
@@ -103,6 +104,8 @@ def get_path_with_params(self, path):
         path_with_params = f"{path}?{urllib.parse.unquote(url_parsed.query)}"
         return path_with_params
 
+    def get_request_query_params(self):
+        return dict(urllib.parse.parse_qsl(urllib.parse.urlsplit(request.url).query))
```
Comment (sergzak022) [inline local_dispatcher/local_dispatch.py:108]: The previous logic that was used to add query params to the request wasn't encoding query params in the url friendly way so it didn't work with minio.  `get_request_query_params` is used to build a dict out of query params so they can be passed as `params` named argument to `requests.requests` method.

File: local_dispatcher/local_dispatch.py [hunk 3]
```diff
@@ -158,11 +162,27 @@ def __init__(self, host=None):
...
-        self.walrus.host = 'http://walrus'     
+        self.walrus.host = 'http://walrus'
         self.frontend.host = 'http://frontend'
         self.frontend.port = 80
 
-        self.app = Flask(__name__, 
+
+        DOCKER_COMPOSE_CONTEXT = env_adapter.bool(os.getenv('DOCKER_COMPOSE_CONTEXT', False))
+
+        if DOCKER_COMPOSE_CONTEXT == True:
```
Comment (sergzak022) [inline local_dispatcher/local_dispatch.py:172]: Add minio proxy if `DOCKER_COMPOSE_CONTEXT` is set.

File: shared/data_tools_core_minio.py [hunk 4]
```diff
@@ -31,8 +31,14 @@ def build_secure_url(self, blob_name: str, expiration_offset: int = None, bucket
         signed_url = super().build_secure_url(blob_name = blob_name,
                                               expiration_offset = expiration_offset,
                                               bucket = bucket)
-        return signed_url
+        if settings.DOCKER_COMPOSE_CONTEXT == True:
+            # note need to proxy request through local dispatcher so users can access from the browser
```
Comment (sergzak022) [inline shared/data_tools_core_minio.py:35]: replace minio url with a url path to a proxy if `DOCKER_COMPOSE_CONTEXT` is true

**Comment by anthony-sarkis** [comment]
Thanks Sergey!

File: local_dispatcher/local_dispatch.py [hunk 5]
```diff
@@ -191,7 +211,7 @@ def start(self):
 
         self.ingress.app.logger.setLevel(logging.INFO)
 
-        self.ingress.app.debug = False
+        self.ingress.app.debug = True
```
Comment (sergzak022) [inline local_dispatcher/local_dispatch.py:214]: @sergzak022 TODO: revert this back to `self.ingress.app.debug = False`

File: docker-compose.yaml [hunk 6]
```diff
@@ -141,6 +142,11 @@ services:
     environment:
       - "MINIO_ROOT_USER=${DIFFGRAM_MINIO_ACCESS_KEY_ID}"
       - "MINIO_ROOT_PASSWORD=${DIFFGRAM_MINIO_ACCESS_KEY_SECRET}"
+    healthcheck:
```
Comment (sergzak022) [inline docker-compose.yaml:145]: `createbuckets` task still fails for me sometimes with `mc: <ERROR> Unable to initialize new alias from the provided credentials. Server not initialized, please try again.` error. I'm looking into making a better healthcheck to try to address it.

File: docker-compose.yaml [hunk 7]
```diff
@@ -11,6 +11,7 @@ services:
     - walrus
     environment:
       DOCKER_CONTEXT: True
+      DOCKER_COMPOSE_CONTEXT: True
```
Comment (sergzak022) [inline docker-compose.yaml:14]: `dispatcher` and `default` need `DOCKER_COMPOSE_CONTEXT` in order to proxy minio asset urls through `dispatcher`

File: docker-compose.yaml [hunk 8]
```diff
@@ -141,12 +142,18 @@ services:
...
+      timeout: 10s
+      retries: 3
     command: server /export --console-address :9001
 
   createbuckets:
     image: minio/mc
     depends_on:
-      - minio
+      minio:
+        condition: service_healthy
```
Comment (sergzak022) [inline docker-compose.yaml:156]: Updated `createbuckets` to wait till minio is healthy.

Comment (anthony-sarkis) [reply docker-compose.yaml:156]: Nice

**Review by anthony-sarkis — APPROVED** [review]

**Pull request merged**